sdks/python/apache_beam/runners/portability/stager.py

```diff
@@ -0,0 +1,573 @@
+    # The file is important so that in the worker the packages are installed
+    # exactly in the order specified. This approach will avoid extra PyPI
+    # requests. For example if package A depends on package B and package A
+    # is installed first then the installer will try to satisfy the
+    # dependency on B by downloading the package from PyPI. If package B is
+    # installed first this is avoided.
+    with open(os.path.join(temp_dir, EXTRA_PACKAGES_FILE), 'wt') as f:
+      for package in local_packages:
+        f.write('%s\n' % os.path.basename(package))
+    staged_path = FileSystems.join(staging_location, EXTRA_PACKAGES_FILE)
+    # Note that the caller of this function is responsible for deleting the
+    # temporary folder where all temp files are created, including this one.
+    self.stage_artifact(
+        os.path.join(temp_dir, EXTRA_PACKAGES_FILE), staged_path)
+    resources.append(EXTRA_PACKAGES_FILE)
+
+    return resources
+
+  def _get_python_executable(self):
+    # Allow overriding the python executable to use for downloading and
+    # installing dependencies, otherwise use the python executable for
+    # the current process.
+    python_bin = os.environ.get('BEAM_PYTHON') or sys.executable
+    if not python_bin:
+      raise ValueError('Could not find Python executable.')
+    return python_bin
+
+  def _populate_requirements_cache(self, requirements_file, cache_dir):
+    # The 'pip download' command will not download again if it finds the
+    # tarball with the proper version already present.
+    # It will get the packages downloaded in the order they are presented in
+    # the requirements file and will not download package dependencies.
+    cmd_args = [
+        self._get_python_executable(),
+        '-m',
+        'pip',
+        'download',
+        '--dest',
+        cache_dir,
+        '-r',
+        requirements_file,
+        # Download from PyPI source distributions.
+        '--no-binary',
+        ':all:'
+    ]
+    logging.info('Executing command: %s', cmd_args)
+    processes.check_call(cmd_args)
+
+  def _build_setup_package(self, setup_file, temp_dir, build_setup_args=None):
+    saved_current_directory = os.getcwd()
+    try:
+      os.chdir(os.path.dirname(setup_file))
+      if build_setup_args is None:
+        build_setup_args = [
+            self._get_python_executable(),
+            os.path.basename(setup_file), 'sdist', '--dist-dir', temp_dir
+        ]
+      logging.info('Executing command: %s', build_setup_args)
+      processes.check_call(build_setup_args)
+      output_files = glob.glob(os.path.join(temp_dir, '*.tar.gz'))
+      if not output_files:
+        raise RuntimeError(
+            'File %s not found.' % os.path.join(temp_dir, '*.tar.gz'))
+      return output_files[0]
+    finally:
+      os.chdir(saved_current_directory)
+
+  def _desired_sdk_filename_in_staging_location(self, sdk_location):
```

Comment:
<!--new_thread; commit:fb22e20e257e1c00a618286cb0a1dd01e4071709; resolved:0-->
Make this a private function in Stager module.